### ISSUE TITLE
3.0 - $this->enableCsrfToken() fail with many requests in the same testAction

### ISSUE THREAD

**Issue by Xety** [cakephp/cakephp#7828]
### 1. Bug with `enableCsrfToken()`
```php
public function testDebug()
{
    $data = [
        'username' => 'mariano',
        'password' => 'password'
    ];
    $this->enableCsrfToken();
    
    //Pass
    $this->post(['controller' => 'users', 'action' => 'login'], $data);
    $this->assertResponseOk();

    //Fail CSRF token doesn't match.
    $this->post(['controller' => 'users', 'action' => 'login'], $data);
    $this->assertResponseOk();
}
```
This is caused by https://github.com/cakephp/cakephp/blob/master/src/TestSuite/IntegrationTestCase.php#L437

#### Solution :
```php
if ($this->_csrfToken === true) {
    $csrfToken = Text::uuid();
    $csrfGenerated = false;

    if (!isset($data['_csrfToken'])) {
        $data['_csrfToken'] = $csrfToken;
        $csrfGenerated = true;
    }
    if (!isset($this->_cookie['csrfToken']) || $csrfGenerated) {
        $this->_cookie['csrfToken'] = $csrfToken;
    }
}
```
And it work well.

### 2. Proposal of a new function `disableCsrfToken()`
Also, it would be nice to have a method `disableCsrfToken()`, because when we call `enableCsrfToken()`, it enable the csrf for all requests in the action.
But if for X or Y reason, we need to disable it for a special request, we can't . (Well, we can with `$this->_csrfToken = false`, but it's more logic to use a function called `disableCsrfToken()`)

Exemple usage :
```php
public function testDebug()
{
    $data = [
        'username' => 'mariano',
        'password' => 'password'
    ];
    $this->enableCsrfToken();
    
    $this->post(['controller' => 'users', 'action' => 'login'], $data);
    $this->assertResponseOk();

    $this->disableCsrfToken();

    $this->post(['controller' => 'users', 'action' => 'login'], $data);
    $this->assertResponseOk();
}
```

**Comment by markstory** [comment]
I generally avoid the need for disablingCsrfToken() by creating multiple test cases. This also prevents state from being shared between test requests.

**Comment by markstory** [comment]
I agree that multiple requests in the same test case method shouldn't cause CSRF checks to fail.

**Comment by dereuromark** [comment]
As a side-note: Never run mulitple requests per test in either case. It can have other side effects and will in many cases fail. Make separate tests per request.

**Comment by Xety** [comment]
@markstory Yeah i understand your point. Anyway, we can still do it with `$this->_csrfToken = false` if we really need it.

@dereuromark Yes, i know it's not good to make several requests in the same action. But sometime, with simple requests, i do it... :stuck_out_tongue_closed_eyes:

**Comment by markstory** [comment]
Multiple request issues fixed with #7839
